pydicom/valuerep.py

```diff
@@ -43,6 +43,318 @@
 # (the component separator '=' is handled separately)
 PN_DELIMS = {0x5e}
 
+# maximum allowed value length for string VRs
+# VRs with a maximum length of 2^32 (UR and UT) are not checked
+MAX_VALUE_LEN = {
+    "AE": 16,
+    "CS": 16,
+    "DS": 16,
+    "IS": 12,
+    "LO": 64,
+    "LT": 10240,
+    "SH": 16,
+    "ST": 1024,
+    "UI": 64
+}
+
+
+def _range_regex(regex: str) -> str:
+    """Compose a regex that allows ranges of the given regex,
+    as defined for VRs DA, DT and TM in PS 3.4, C.2.2.2.5.
+    """
+    return fr"^{regex}$|^\-{regex} ?$|^{regex}\- ?$|^{regex}\-{regex} ?$"
+
+
+# regular expressions to match valid values for some VRs
+AE_REGEX = re.compile(r"^[\x20-\x7e]*$")
+AS_REGEX = re.compile(r"^\d\d\d[DWMY]$")
+CS_REGEX = re.compile(r"^[A-Z0-9 _]*$")
+DS_REGEX = re.compile(
+    r"^ *[+\-]?(\d+|\d+\.\d*|\.\d+)([eE][+\-]?\d+)? *$"
+)
+IS_REGEX = re.compile(r"^ *[+\-]?\d+ *$")
+DA_REGEX = re.compile(_range_regex(r"\d{4}(0[1-9]|1[0-2])([0-2]\d|3[01])"))
+DT_REGEX = re.compile(_range_regex(
+    r"\d{4}((0[1-9]|1[0-2])(([0-2]\d|3[01])(([01]\d|2[0-3])"
+    r"([0-5]\d((60|[0-5]\d)(\.\d{1,6} ?)?)?)?)?)?)?([+-][01]\d\d\d)?")
+)
+TM_REGEX = re.compile(_range_regex(
+    r"([01]\d|2[0-3])([0-5]\d((60|[0-5]\d)(\.\d{1,6} ?)?)?)?"))
+UR_REGEX = re.compile(r"^[A-Za-z_\d:/?#\[\]@!$&'()*+,;=%\-.~]* *$")
+
+
+def validate_vr_length(vr: str, value: Union[str, bytes]) -> Tuple[bool, str]:
+    """Validate the value length for a given VR.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str or bytes
+        The value to validate.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    max_length = MAX_VALUE_LEN.get(vr, 0)
+    if max_length > 0:
+        value_length = len(value)
+        if value_length > max_length:
+            return False, (
+                f"The value length ({value_length}) exceeds the "
+                f"maximum length of {max_length} allowed for VR {vr}."
+            )
+    return True, ""
+
+
+def validate_regex(vr: str, value: Union[str, bytes],
```

Comment:
I was just testing this out, not sure how important this is but if you pass in a byte-string you get a TypeError, should this be changed to `value: str`?

Ex.:

```python
➜  /w/f/s/pydicom git:(master) ✗ $ python
Python 3.8.8 (default, Oct 11 2021, 09:15:09)
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pydicom.valuerep import validate_valu
>>> validate_value('DS', b"3000.0",1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/work/flywheel/sse/pydicom/pydicom/valuerep.py", line 342, in validate_value
    is_valid, msg = validator(vr, value)
  File "/work/flywheel/sse/pydicom/pydicom/valuerep.py", line 181, in validate_ds
    return validate_length_and_regex(vr, value, DS_REGEX)
  File "/work/flywheel/sse/pydicom/pydicom/valuerep.py", line 143, in validate_length_and_regex
    is_valid_expr, msg2 = validate_regex(vr, value, regex)
  File "/work/flywheel/sse/pydicom/pydicom/valuerep.py", line 119, in validate_regex
    if value and (not re.match(regex, value) or value and value[-1] == "\n"):
  File "/home/naterichman/.pyenv/versions/3.8.8/lib/python3.8/re.py", line 191, in match
    return _compile(pattern, flags).match(string)
TypeError: cannot use a string pattern on a bytes-like object
```

Reply:
Yea I should have clarified, this isn't a problem with the normal pydicom flow.  But I saw this and assumed I could pass a byte string, more just user-error or needing docs.

Reply:
I think it's fair to require `convert_values` to be called beforehand, then it could be only str inputs

Reply:
Ok, in this case this should probably be documented.

Reply:
Actually I think I will change this to support both `str` and `byte`, as for some values it has to support `bytes`.
@darcymason : I would like to change this before the release, if that is ok.

Reply:
Missed replying to this one - all good since merged now anyway